### PR TITLE
[New feature] integrate causal_conv1d Triton kernel for Ascend NPU 

### PR DESCRIPTION
**Pull request by ys2025-AI**
# PR type
- [ ] Bug Fix
- [x] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information

integrate causal_conv1d Triton kernel for Ascend NPU

## Experiment results
Model: Qwen3.5-4B
Hardware: Atlas 900 A3 (2 x NPU)
Dataset: GSM8K_ZH
Finetuning type: LoRA
Software: cann9.0.0+ torch/orch_npu 2.9.0 + MindSpeed 0.12.1 + triton-ascend 3.2.1 + transformers 5.9
| 指标                | Baseline      | Causal\_conv1d优化 | 差异                |
| ----------------- | ------------- | ---------------- | ----------------- |
| **加速比**           | 1.0x          | **1.12x**        | —                 |
| **平均 loss**       | 0.6449        | 0.6456           | 差异 0.0007         |

related: https://gitcode.com/Ascend/MindSpeed-Ops

### CODE REVIEW

File: src/twinkle/kernel/causal_conv1d.py [hunk 1]
```diff
@@ -0,0 +1,1103 @@
...
+
+                    b_dw_sub = tl.sum(b_dy_sub * b_x, 0)
+                    b_dw = tl.insert_slice(b_dw, b_dw_sub[None, :], [W - i_w - 1, 0], [1, BD], [1, 1])
+
+                if HAS_BIAS and i_w == 0:
+                    b_db += tl.sum(b_dy_sub, 0)
+                b_dx += b_wdy
+
+            p_dw = tl.make_block_ptr(dw + i_tg * W * D, (W, D), (D, 1), (0, i_d * BD), (W, BD), (1, 0))
+            tl.store(p_dw, b_dw.to(dw.dtype.element_ty))
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply src/twinkle/kernel/causal_conv1d.py]: Added the HAS_WEIGHT conditional guard

File: src/twinkle/patch/gdn_padding_free.py [hunk 2]
```diff
@@ -49,25 +54,51 @@ def _patch_gdn_kernels_for_cu_seqlens(
...
+        def causal_conv1d_wrapper(*args, **kwargs):
+            x = kwargs.pop('x')
+            del kwargs['seq_idx']
+            del kwargs['backend']
+            y, _ = ms_causal_conv1d(
+                x=x,
+                cu_seqlens=cu_seqlens.to(dtype=torch.int32, device=x.device),
+                **kwargs,
+            )
+            return y
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply src/twinkle/patch/gdn_padding_free.py]: FLA passes channel-first [B, D, T] inputs, but the NPU kernel expects [B, T, D] inputs and [W, D] weights. The original implementation did not transpose and completely dropped *args, causing a guaranteed shape mismatch and missing arguments. Now aligned with the CUDA path's transpose logic.

File: src/twinkle/kernel/causal_conv1d.py [hunk 3]
```diff
@@ -0,0 +1,1103 @@
...
+
+
+@functools.lru_cache(maxsize=1)
+def _prepare_lens(cu_seqlens: torch.LongTensor) -> torch.LongTensor:
+    return cu_seqlens[1:] - cu_seqlens[:-1]
+
+
+def _prepare_chunk_indices(cu_seqlens: torch.LongTensor, chunk_size: int) -> torch.LongTensor:
+    indices = torch.cat([torch.arange(n) for n in triton.cdiv(_prepare_lens(cu_seqlens), chunk_size).tolist()])
+    return torch.stack([indices.eq(0).cumsum(0) - 1, indices], 1).to(cu_seqlens)
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply src/twinkle/kernel/causal_conv1d.py]: fixed

File: src/twinkle/kernel/causal_conv1d.py [hunk 4]
```diff
@@ -0,0 +1,1103 @@
...
+    if x.shape[-1] != weight.shape[-1]:
+        raise ValueError('x [B, T, D], weight [W, D], please check.')
+    B, T, D, W = *x.shape, weight.shape[0]
+    NUM_CORES = _get_vector_num()
+    if initial_state is not None:
+        BD = 32
+        BT = min(16, triton.next_power_of_2(triton.cdiv(max(16, B * T), NUM_CORES)))
+    else:
+        BD = 256
+        BT = min(32, triton.next_power_of_2(triton.cdiv(max(16, B * T), NUM_CORES)))
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply src/twinkle/kernel/causal_conv1d.py]: Dynamic BD computation ensures compatibility with common model dimensions.

File: src/twinkle/kernel/causal_conv1d.py [hunk 5]
```diff
@@ -0,0 +1,1103 @@
...
+                nst_mask = (nst_off_y0 < dim) & (nst_off_y1 >= 0) & (nst_off_y1 < state_len)
+                block_ptr = bi * dim * state_len + nst_off_y0 * state_len + nst_off_y1
+                tl.store(conv_state_update_ptr + block_ptr, x_new_s, mask=nst_mask)
+
+        for owi in tl.range(0, width):
+            new_x = tl.extract_slice(x_b, (owi, 0), (T_CHK_SIZE, D_CHK_SIZE), (1, 1))
+            w_chl_wi = tl.extract_slice(w, (owi, 0), (1, D_CHK_SIZE), (1, 1))
+            x_mul_chl_wi = new_x * w_chl_wi
+            out_block += x_mul_chl_wi
+        out_block = tl.trans(out_block, (1, 0))
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply src/twinkle/kernel/causal_conv1d.py:916]: Added bias loading and accumulation.

**Review by tpx818 — APPROVED** [review]
